subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java

```diff
@@ -131,64 +172,68 @@ public DefaultJvmTestSuite(String name, DependencyFactory dependencyFactory, Con
                         default:
                             throw new IllegalStateException("do not know how to handle " + vtf);
                     }
-                }));
+                    // In order to maintain compatibility for the default test suite, we need to load JUnit4 from the Gradle distribution
+                    // instead of including it in testImplementation.
+                    // TODO: In 8.0 should we switch getVersionedTestingFramework() to use JUNIT4 as a convention?
+                }).orElse(new DefaultProvider<>(() -> frameworkLookup.computeIfAbsent(null, f -> new JUnitTestFramework(task, (DefaultTestFilter) task.getFilter(), true)))));
             });
         });
     }
 
-    private void attachDependenciesForTestFramework(DependencyFactory dependencyFactory, DependencyAdder implementation) {
+    private void setFrameworkTo(TestingFramework framework, Provider<String> version) {
+        getVersionedTestingFramework().set(version.map(v -> new VersionedTestingFramework(framework, v)));
+
+        // This whole way of adding the implementation dependencies here is messed up. Once a user calls the useXXX method,
+        // they can't go back. Maybe that's no so terrible, except that we call useJunitJupiter() FOR ALL USER DEFINED TEST SUITES.
+        // This is a real problem. Users who want to use anything but JUnit Jupiter will always have the JUnit Jupiter
+        // on their classpath and will have to manually declare the implementation dependencies for the test framework
+        // they want to use.
+
+        // There is a nice-seeming solution: We could move the line below, `.getImplementation.bundle(...)`, into the constructor.
+        // This could seemingly work, except for if the configuration's dependencies ever get resolved early, before the user is
+        // able to call their useXXX method. If that happens, the provider gets resolved early, and the dependencies defined at
+        // that point are added to the configuration. Then, after the user calls their useXXX method, the provider is updated
+        // but the configuration never is. This is a problem with how we declare dependencies on configurations. We should be
+        // able to restrict users from resolving the configuration dependencies early OR make these dependencies even more lazy
+        // so that if dependencies are resolved early, the future changes in the lazy properties are still propagated to later resolutions.
+
+        // A concrete example of this occurs when using the Kotlin Gradle Plugin. See the function `configureKotlinTestDependency` added
+        // at the below linked commit. This resolves the dependencies before configuration-time is over, and would break an implementation
+        // where the below line is instead located in the constructor.
+        // See: https://github.com/JetBrains/kotlin/commit/4a172286217a1a7d4e7a7f0eb6a0bc53ebf56515
```

Comment:
Is there any way to use `Configuration.beforeLocking(Action)` to register a callback to check if the dependencies are being resolved too soon and add a warning in that case?  Then we could provide a warning and change this in Gradle 9.0.

Reply:
Looking into this further is definitely on my to-do list. I've already looked deep into the provider code and `DomainObjectCollection` implementation. The root of the problem is the DomainObjectCollection caching providers -- even when it can't guarantee that the provider won't change in the future. There is some code built-in to handle some special providers which do declare it is changing, though that code is very rudimentary. I think a good bit of work might be required to look into this, and I think any more work on this PR related to this issue is out of scope. I've spent more than enough time here and we should work this issue separately. 

As to your recommendation specifically, checking if the dependencies have been resolved early would seemingly require a bit too much mucking around in `DomainObjectCollection` code for me to want to include it in this PR. As of right now, the way this works in this PR as compared to before has not changed, it is just better documented now. 

I do want to follow up on this shortly though. The sooner we can fix this the cleaner we can make all this code.